added proposal for authenticity handling #23

svenkubiak opened the pull request:
This is a proposal for authenticity handling from #23 

You call in the template either ${authenticity("form")} or ${authenticity("token")} and it is checked via the filter, when you add the filter to one of your controller methods.

This proposal is open for discussion. I will add tests and update the documentation once we agree on a final way to go.

Commit:
added proposel for authenticity handling #23

buildhive commented:
[Ninja Web Framework » ninja #438](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/438/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

gitblit commented:
That's a very good idea to create a new template engine method to cut down on boilerplate. :+1:

svenkubiak commented:
Like I said. It's open for discussion.

raphaelbauer commented on ninja-core/src/main/java/ninja/template/TemplateEngineFreemarker.java:
```diff
@@ -263,6 +263,8 @@ public void invoke(Context context, Result result) {
         map.put("assetsAt", templateEngineFreemarkerAssetsAtMethod);
         map.put("webJarsAt", templateEngineFreemarkerWebJarsAtMethod);
+        
+        map.put("authenticity", new TemplateEngineFreemarkerAuthenticityMethod(context.getSession().getAuthenticityToken()));
```
Just pass the context. Right now you force the creation of the session cookie and authenticity token for requests that potentially don't use them. The call to getSession().getAuthenticityToken() should be executed only when the template method is called...

raphaelbauer commented:
I think your solution is pretty nice :)

Commit:
updated documentation

Commit:
minor refactoring

svenkubiak commented:
@raphaelbauer 
I changed the code according to your comment and added the documentation. Tests are coming soon.

I am note sold on the template calls. What would be pretty cool, would be a global macro and just call

```
@<authenticityToken/>
@<authenticityForm/>
```

But to make this work flawless, this would has to be available in the freemarker global namespace and to be made available programmatically. Does anyone know if this is possible at all?

buildhive commented:
[Ninja Web Framework » ninja #439](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/439/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #440](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/440/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

Commit:
added unit tests

buildhive commented:
[Ninja Web Framework » ninja #441](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/441/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #442](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/442/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

raphaelbauer commented:
Simplifying the syntax via a macro would be a cool idea. You can ask the Freemarker community at http://freemarker.org/mailing-lists.html . They are really helpful and friendly :)

svenkubiak commented:
Ok. Lets wait for a reply on stackoverflow http://stackoverflow.com/questions/28472751/is-it-possible-to-create-a-freemarker-macro-programmatically

svenkubiak commented:
No solid answer on stackoverflow yet. If the PR is okay, I suggest we'll go with it for and maybe do a refactoring afterwards.

t3hc13h commented:
Couldn't you get the desired effect by registering a TemplateDirectiveModel as a shared variable in the template configuration?

svenkubiak commented:
@t3hc13h I did try the TemplateDirectiveModel, but the passed TemplateDirectiveBody is null as I just call 

   <@myDirective/>

Commit:
refactored to TemplateDirectiveModel

buildhive commented:
[Ninja Web Framework » ninja #455](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/455/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

svenkubiak commented:
Ok, I figured it finally out. You can now just call

```
<@authenticityToken/>
<@authenticityForm/>
```

@raphaelbauer 
I updated the Pull-Request.

raphaelbauer commented on ninja-core/src/test/java/ninja/AuthenticityFilterTest.java:
```diff
@@ -0,0 +1,73 @@
+        when(context.getParameter(NinjaConstant.AUTHENTICITY_TOKEN)).thenReturn("foo");
+        when(context.getSession()).thenReturn(session);
+        when(context.getSession().getAuthenticityToken()).thenReturn("bar");
+        when(ninjaDefault.getForbiddenResult(context)).thenReturn(forbidden);
```
you can simplify the code a bit... we don't need ok and forbidden results...

```
Just do a check  Mockito.verify(ninjaDefault).getForbiddenResult(context)
```

or

```
Mockito.verify(filterChain).next(context);
```

assertNotNull and assertEquals is not needed imho...

That makes the code a bit shorter...

raphaelbauer commented on ninja-core/src/site/markdown/documentation/html_templating/advanced_topics.md:
```diff
@@ -133,6 +133,19 @@ prettyTime allows you to format localized relative dates.
 For instance, if you had a date object that represented yesterday, prettyTime would format that as *1 day ago* in the
 preferred Locale of the request.
 
+### authenticity(...)
```
very cool...

raphaelbauer commented on ninja-core/src/main/java/ninja/template/directives/TemplateEngineFreemarkerAuthenticityTokenDirective.java:
```diff
@@ -0,0 +1,40 @@
+ *
+ */
+@SuppressWarnings("rawtypes")
+public class TemplateEngineFreemarkerAuthenticityTokenDirective implements TemplateDirectiveModel {
```
we need a testcase for the directive...

raphaelbauer commented on ninja-core/src/main/java/ninja/template/directives/TemplateEngineFreemarkerAuthenticityFormDirective.java:
```diff
@@ -0,0 +1,42 @@
+ *
+ */
+@SuppressWarnings("rawtypes")
+public class TemplateEngineFreemarkerAuthenticityFormDirective implements TemplateDirectiveModel {
```
we need a testcase for the directive...

raphaelbauer commented on ninja-core/src/main/java/ninja/template/TemplateEngineFreemarker.java:
```diff
@@ -263,6 +264,9 @@ public void invoke(Context context, Result result) {
         map.put("assetsAt", templateEngineFreemarkerAssetsAtMethod);
         map.put("webJarsAt", templateEngineFreemarkerWebJarsAtMethod);
+        
+        map.put("authenticityToken", new TemplateEngineFreemarkerAuthenticityTokenDirective(context));
```
we need a test to make sure authenticityToken and authenticityForm is added to the templating engine... We should really make sure that your great feature does not get destroyed by eg a bogus merge...

You can do that either via a unit test or in the ninja-integration-test project...

raphaelbauer replied:
Nevermind - I'll do that locally while merging...

raphaelbauer commented:
Hey Sven - the feature, implementation and documentation is pretty cool. I added some remarks. I think we need more test-cases. Otherwise your great feature may get lost between some bogus merges for instance... This really helps to produce stable software in the long run which is good for all of us :) Hope that makes sense - let me know if I can help!

Commit:
added test cases for authenticity directives

buildhive commented:
[Ninja Web Framework » ninja #460](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/460/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

Commit:
updated test case for authenticity filter

buildhive commented:
[Ninja Web Framework » ninja #461](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/461/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #462](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/462/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

Commit:
added integration test for authenticity checks

buildhive commented:
[Ninja Web Framework » ninja #463](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/463/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #464](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/464/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

Commit:
fixed merge conflicts in changelog

buildhive commented:
[Ninja Web Framework » ninja #465](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/465/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #466](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/466/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

Commit:
fixed merge conflicts in changelog (2)

Commit:
updated changelog

Commit:
fixed merge conflicts in changelog (3)

svenkubiak commented:
@raphaelbauer 
I had some problems with the test cases, but finally figured it out what the problem was. Should be all good now :-D

Could you help me on the merge conflicts by adding the changelog entry after merging?

buildhive commented:
[Ninja Web Framework » ninja #467](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/467/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #468](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/468/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

raphaelbauer commented on ninja-core/src/main/java/ninja/template/directives/TemplateEngineFreemarkerAuthenticityTokenDirective.java:
```diff
@@ -0,0 +1,40 @@
+    private String authenticityToken;
+
+    public TemplateEngineFreemarkerAuthenticityTokenDirective(Context context) {
+        this.authenticityToken = context.getSession().getAuthenticityToken();
```
please do this in the execute method. Otherwise you are always adding the __AT to the session - even if it is not needed...

raphaelbauer commented on ninja-core/src/main/java/ninja/template/directives/TemplateEngineFreemarkerAuthenticityFormDirective.java:
```diff
@@ -0,0 +1,42 @@
+    private String authenticityToken;
+
+    public TemplateEngineFreemarkerAuthenticityFormDirective(Context context) {
+        this.authenticityToken = context.getSession().getAuthenticityToken();
```
please do this in the execute method. Otherwise you are always adding the __AT to the session - even if it is not needed...

raphaelbauer replied:
I'll do that locally...

svenkubiak commented:
Ok, Just let me know, if there is anything left to do ;-)

Commit:
change call to authenticity token

svenkubiak commented:
@raphaelbauer 
Was a quick fix ;-)

buildhive commented:
[Ninja Web Framework » ninja #476](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/476/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #477](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/477/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

raphaelbauer commented:
I just changed the test-cases a bit. Really good work Sven - great feature - great addition to Ninja!
